middleware/src/hsm/hsm.go

```diff
@@ -203,3 +204,18 @@ func (hsm *HSM) InteractWithBootloader(f func(*bb02bootloader.Device)) error {
 	f(device)
 	return nil
 }
+
+// UpdateFirmware reboots into Bootloader and executes the firmware update
+func (hsm *HSM) UpdateFirmware(hsmFirmwareFile string) error {
+	hsmFirmwareBinary, err := ioutil.ReadFile(hsmFirmwareFile)
+	if err != nil {
+		return err
+	}
+	err = hsm.InteractWithBootloader(func(bootloader *bb02bootloader.Device) {
+		err = bootloader.UpgradeFirmware(hsmFirmwareBinary)
```

Comment:
it seems worthwhile to compare the monotonic version here again, as if the HSM boots in the bootloader from the start, the version is still checked.

alternatively, booting in booloader mode means something went wrong, so you could also just upgrade by default in that case.

as discussed offline, if this call fails, probably it's good to reboot the base and try again.

Reply:
Addressed in the following way:

initHSM tries to connect to the HSM firmware. On success, it checks if a firmware upgrade is available and tries to apply it.
On failing to connect to the firmware, it tries to connect to the bootloader and re-install the firmware.
If that also fails, we continue without the HSM

On purpose I am not rebooting automatically if any of this fails, as I don't want the Base to get stuck in a loop of failed upgrades. If a user see the upgrade has failed, they can reboot/re-plug manually.

If we decide to handle this automatically, we should store a persistent counter of failed upgrades and only try to reflash 2 or 3 times before informing the user that something has gone wrong.